### PR TITLE
kubectl proxy: append context host path to request path

### LINKED ISSUE #94280: kubectl proxy not working for kube contexts with base path
**What happened**:

Given that you have a kube context that looks like this:

```yaml
apiVersion: v1
kind: Config
clusters:
- cluster:
    server: https://my-url-to-kubernetes.com/any-base-path-to-kubernetes
  name: default
contexts:
- context:
    cluster: default
    user: default
  name: default
current-context: default
users:
- name: default
  user:
    client-certificate-data: ANY
    client-key-data: ANY
```

and you try to run `kubectl proxy`, the request `http://localhost:8081/version` fails, because internally the URL is not correctly rewritten to `https://my-url-to-kubernetes.com/any-base-path-to-kubernetes/version` and instead calls `https://my-url-to-kubernetes.com/version`. The problem is in file (at least what I think) https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L231 , where the path needs to be passed as well (also for upgrade connections at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L273)

You can see the problem via `kubectl proxy --v 10`:

```
I0827 14:18:25.042177   30678 loader.go:375] Config loaded from file:  /Users/fabiankramm/.kube/config
Starting to serve on 127.0.0.1:8001
I0827 14:18:28.414766   30678 proxy_server.go:88] /version matched ^.*
I0827 14:18:28.415391   30678 proxy_server.go:88] 127.0.0.1 matched ^127\.0\.0\.1$
I0827 14:18:28.415610   30678 proxy_server.go:128] Filter accepting GET /version 127.0.0.1
I0827 14:18:28.416002   30678 upgradeaware.go:253] Request was not an upgrade
I0827 14:18:28.416841   30678 round_trippers.go:423] curl -k -v -XGET  -H "Cache-Control: max-age=0" -H "Accept-Language: de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7" -H "Sec-Fetch-Mode: navigate" -H "User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36" -H "Accept-Encoding: gzip, deflate, br" -H "X-Forwarded-For: 127.0.0.1" -H "Sec-Fetch-Site: none" -H "Sec-Fetch-User: ?1" -H "Sec-Fetch-Dest: document" -H "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9" -H "Upgrade-Insecure-Requests: 1" 'https://test.localhost/version'
I0827 14:18:28.654599   30678 round_trippers.go:443] GET https://test.localhost/version 404 Not Found in 237 milliseconds
I0827 14:18:28.654637   30678 round_trippers.go:449] Response Headers:
I0827 14:18:28.654644   30678 round_trippers.go:452]     Date: Thu, 27 Aug 2020 12:18:29 GMT
I0827 14:18:28.654649   30678 round_trippers.go:452]     Content-Type: text/plain; charset=utf-8
I0827 14:18:28.654654   30678 round_trippers.go:452]     Content-Length: 21
I0827 14:18:28.654659   30678 round_trippers.go:452]     Server: nginx/1.19.1
```

As you can see the request is made to 'https://test.localhost/version', even though the kube config has a base path of '/kubernetes':

```yaml
apiVersion: v1
kind: Config
clusters:
- cluster:
    certificate-authority-data: ANY
    server: https://test.localhost/kubernetes
  name: default
contexts:
- context:
    cluster: default
    user: default
  name: default
current-context: default
users:
- name: default
  user:
    client-certificate-data: ANY
    client-key-data: ANY
```

**What you expected to happen**:

The request `http://localhost:8081/version` works, when using `kubectl proxy`

**How to reproduce it (as minimally and precisely as possible)**:

- Spin up a cluster with api root at a different base path like /kubernetes
- Run `kubectl proxy` 
- Try `http://localhost:8081/version` in the webbrowser

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
```
minikube version: v1.12.3
commit: 2243b4b97c131e3244c5f014faedca0d846599f5-dirty
```

### PR DESCRIPTION
**Pull request by FabianKramm**
Signed-off-by: fabiankramm <fab.kramm@googlemail.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR adds a new flag `--append-server-path` to `kubectl proxy` that changes the way how kube contexts are handled that specify a server with an appended path, e.g. `https://my-apiserver.com/path/to/kubernetes`. Before this change, `kubectl proxy` would always expose the api at `http://127.0.0.1:8001/path/to/kubernetes/...`, with this change and flag enabled it exposes the api at `http://127.0.0.1:8001/...` and automatically appends the kube context server path. 

**Which issue(s) this PR fixes**:

Fixes #94280

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Added a new flag `--append-server-path` to `kubectl proxy` that will automatically append the kube context server path to each request.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
none
```

### CODE REVIEW

**Comment by FabianKramm** [comment]
@pwittrock @eddiezane

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go [hunk 1]
```diff
@@ -60,6 +60,8 @@ type UpgradeAwareHandler struct {
 	// Location is the location of the upstream proxy. It is used as the location to Dial on the upstream server
 	// for upgrade requests unless UseRequestLocationOnUpgrade is true.
 	Location *url.URL
+	// PrependLocationPath determines if the original path of the Location should be prepended to the upstream proxy request path
+	PrependLocationPath bool
```
Comment (eddiezane) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:64]: I think we want `AppendLocationPath`. Or maybe `UseLocationPath`?

Anyone have thoughts?

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go [hunk 2]
```diff
@@ -238,6 +246,12 @@ func (h *UpgradeAwareHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
 		// or supporting retry mechanisms by not sending non-fatal errors to the clients
 		proxy.ErrorHandler = h.Responder.Error
 	}
+
+	// exchanging req.Host with the upstream proxy location is necessary for HTTP/2 backends, because req.Host
+	// has preference over req.URL.Host there. Unfortunately, httputil.ReverseProxy does not set req.Host
+	// and only sets req.URL.Host in the request, which can lead to an endless loop, since the request is rerouted to the same location.
+	// (see https://github.com/golang/go/issues/28168 for more information)
+	newReq.Host = reverseProxyLocation.Host
```
Comment (eddiezane) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:254]: Should this be in a separate PR?

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go [hunk 3]
```diff
@@ -222,6 +223,13 @@ func TestServeHTTP(t *testing.T) {
 				"Access-Control-Allow-Methods",
 			},
 		},
+		{
```
Comment (eddiezane) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go:226]: I haven't looked at the other tests but can we add a test that asserts this isn't followed when not set?

**Review by eddiezane — COMMENTED** [review]
Thanks for opening this PR!

I think we should consider this a breaking change and make it opt-in instead of opt-out.

We could deprecate the old behavior? 

Thoughts @soltysh?

**Comment by lavalamp** [comment]
I won't be able to review right now (also this is SIG CLI) but I think you should make it opt-in rather than opt-out.

**Comment by FabianKramm** [comment]
@eddiezane @lavalamp thanks for taking a look! Regarding the opt-in vs opt-out, currently I think this PR is more a bug fix for most tools that use `kubectl proxy` internally (e.g. [Lens](https://github.com/lensapp/lens)), because I would assume 99% of them don't know about this issue and just assume the api server is always reachable directly under `/` for every kube context. In addition, the documentation of the command never mentions this exception and it seems like the api is always directly reachable under `/`, so I think you could argue that this is more a bug fix than a breaking change.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go (hunk 1 above)
Comment (FabianKramm) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go]: Changed it to `AppendLocationPath`

**Comment by FabianKramm** [comment]
@eddiezane I changed the things that you reviewed and made it opt-in.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go (hunk 2 above)
Comment (FabianKramm) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go]: I added a new PR #97385 for this

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go (hunk 3 above)
Comment (FabianKramm) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go:226]: The other tests were already implicitly testing this, however I added another test to make this explicit

**Comment by FabianKramm** [comment]
@soltysh @dougsland any updates on this?

File: staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go [hunk 4]
```diff
@@ -173,8 +173,8 @@ func makeUpgradeTransport(config *rest.Config, keepalive time.Duration) (proxy.U
 
 // NewServer creates and installs a new Server.
 // 'filter', if non-nil, protects requests to the api only.
-func NewServer(filebase string, apiProxyPrefix string, staticPrefix string, filter *FilterServer, cfg *rest.Config, keepalive time.Duration) (*Server, error) {
-	proxyHandler, err := NewProxyHandler(apiProxyPrefix, filter, cfg, keepalive)
+func NewServer(filebase string, apiProxyPrefix string, staticPrefix string, filter *FilterServer, cfg *rest.Config, keepalive time.Duration, appendLocationPath bool) (*Server, error) {
```
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go:176]: @lavalamp How bad is change the NewServer function signature to add a new option at this stage?

File: staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go [hunk 5]
```diff
@@ -189,7 +189,7 @@ func NewServer(filebase string, apiProxyPrefix string, staticPrefix string, filt
 }
 
 // NewProxyHandler creates an api proxy handler for the cluster
-func NewProxyHandler(apiProxyPrefix string, filter *FilterServer, cfg *rest.Config, keepalive time.Duration) (http.Handler, error) {
+func NewProxyHandler(apiProxyPrefix string, filter *FilterServer, cfg *rest.Config, keepalive time.Duration, appendLocationPath bool) (http.Handler, error) {
```
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go:192]: @lavalamp the same apply here.

**Review by dougsland — COMMENTED** [review]
@FabianKramm sorry the delay, few comments there.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go [hunk 6]
```diff
@@ -400,6 +411,19 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 	return true
 }
 
+// this method is taken from net/http/httputil/reverseproxy.go
```
[the remark this answers is not in the document]
Comment (FabianKramm) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:414]: @dougsland the function unfortunately isn't exported in https://golang.org/src/net/http/httputil/reverseproxy.go . Maybe I'm missing something, but I don't see a way to use it directly here.

File: staging/src/k8s.io/kubectl/pkg/cmd/proxy/proxy.go [hunk 7]
```diff
@@ -37,18 +37,19 @@ import (
...
-	acceptPaths   string
-	rejectPaths   string
-	acceptHosts   string
-	rejectMethods string
-	port          int
-	address       string
-	disableFilter bool
-	unixSocket    string
-	keepalive     time.Duration
+	staticDir        string
```
[the remark this answers is not in the document]
Comment (FabianKramm) [reply staging/src/k8s.io/kubectl/pkg/cmd/proxy/proxy.go:40]: @dougsland sure, I can move appendServerPath to a new line and leave the indentation for the rest

File: staging/src/k8s.io/kubectl/pkg/proxy/proxy_server_test.go [hunk 8]
```diff
@@ -452,7 +452,7 @@ func TestPathHandling(t *testing.T) {
 
 	for _, tt := range table {
 		t.Run(tt.name, func(t *testing.T) {
-			p, err := NewServer("", tt.prefix, "/not/used/for/this/test", nil, cc, 0)
+			p, err := NewServer("", tt.prefix, "/not/used/for/this/test", nil, cc, 0, false)
```
[the remark this answers is not in the document]
Comment (FabianKramm) [reply staging/src/k8s.io/kubectl/pkg/proxy/proxy_server_test.go:455]: Sure, I can add a test with this being enabled

[the remark this answers is not in the document]
Comment (FabianKramm) [reply staging/src/k8s.io/kubectl/pkg/proxy/proxy_server_test.go]: @dougsland done

File: staging/src/k8s.io/kubectl/pkg/cmd/proxy/proxy.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (FabianKramm) [reply staging/src/k8s.io/kubectl/pkg/cmd/proxy/proxy.go]: @dougsland done

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dougsland) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:414]: > @dougsland the function unfortunately isn't exported in https://golang.org/src/net/http/httputil/reverseproxy.go .
> Maybe I'm missing something, but I don't see a way to use it directly here.

You are correct. I would suggest to share that in the comment too. Additionally (if possible), reach out golang folks to may export the function to avoid duplication code and others projects using it.

For instance:
// FIXME: Taken from net/http/httputil/reverseproxy.go as singleJoiningSlash is not exported to be re-used.  
// See-also: https://github.com/golang/issues/1234567

**Review by dougsland — CHANGES_REQUESTED** [review]

**Comment by dougsland** [comment]
Adding more people for review
/cc @rikatz

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (FabianKramm) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go]: @dougsland changed the comment and opened an issue in the golang repository: https://github.com/golang/go/issues/44290

**Comment by FabianKramm** [comment]
@lavalamp do you think it might be possible that you take a quick look at the comments of @dougsland? Since my other PR #97385 got merged now, I rebased this one to resolve the merge conflicts.

File: staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go (hunk 4 above)
Comment (lavalamp) [reply staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go:176]: I don't know what compatibility guarantees SIG CLI offers on kubectl code-- I wouldn't expect any but I don't know.

Comment (dougsland) [reply staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go:176]: @lavalamp Thanks, let me put into the thread @soltysh. Maciej, could you please review this PR and share your thoughts please?

**Review by soltysh — COMMENTED** [review]
In the original issue #94280 it was mentioned it's solving a special use-case, before we decide to ship it I'd like to know how many those special cases there are, I don't want to grow both our code-base and our flags to solve all those tiny corner cases. We in SIG-CLI are working on documenting a process which will allow us to better decide about future requests, and I'm seeing this as just that. It's not a bug, but rather a conscious limitation of the functionality.

**Comment by FabianKramm** [comment]
@soltysh thanks a lot for your review! Appending a base path in a kube context is a common practice for kubernetes api gateways that route requests to multiple different kubernetes clusters under a common domain, for example products like [rancher](https://rancher.com/docs/rancher/v2.x/en/overview/architecture/) or [loft](https://loft.sh/docs/architecture) (thats where I'm working) are doing this. A product that builds upon `kubectl proxy` is for example [lens](https://github.com/lensapp/lens). Here are some issues from rancher and lens regarding this problem and we also had several users complain that lens is not working correctly with loft because of this:
- https://github.com/rancher/docs/issues/929 (basically says you shouldn't use `kubectl proxy` with rancher at all, they also referenced the issue here)
- https://github.com/rancher/rancher/issues/13979 (same issue here)
- https://github.com/lensapp/lens/issues/91#issuecomment-559471143 (same issue here but they implemented a workaround)

Currently I think the default behaviour of `kubectl proxy` is really unexpected if your kube context has a base path (since there is no documentation or any mention of this special case), but I also fully agree with you on the growing code base and flag forest concern. Thats why I really like what @eddiezane has proposed that we deprecate the old behaviour if this gets merged and then sooner or later change the default behaviour and ultimately remove the flag again. Do you think that is a possibility?

**Comment by FabianKramm** [comment]
@soltysh Regarding other special cases in `kubectl proxy`: I can't really judge this, I personally only know about this issue, because we had several users complain about it and for them parsing the kube config themselves to find out the base path to append is not an acceptable option (basically what the rancher folks are saying as well). So I would be really glad if this would make it one way or another into kubectl itself.

**Comment by FabianKramm** [comment]
@soltysh @eddiezane @dougsland I just wanted to ping again, if there is any possibility that we can get this merged as more people having problems with this (see https://github.com/operator-framework/operator-sdk/issues/4925 for example). I really believe this is a needed flag and `kubectl` itself should ship a solution for this.

**Comment by varshaprasad96** [comment]
@soltysh @eddiezane @dougsland any updates on when this PR would be merged?

**Comment by dougsland** [comment]
Hi,

> @soltysh @eddiezane @dougsland any updates on when this PR would be merged?

@varshaprasad96 @FabianKramm I would suggest raise this PR in SIG-CLI meeting to get attention from the group.

Here the data for the meeting: 
Regular SIG Meeting: Wednesdays at 09:00 PT (Pacific Time) (biweekly).
https://github.com/kubernetes/community/tree/master/sig-cli

File: staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go (hunk 4 above)
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go:176]: Yeah, kubectl-wise I'm ok with extending this method signature.

File: staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go (hunk 5 above)
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go:192]: Ditto.

**Review by soltysh — APPROVED** [review]
I'm fine with this as is, the only addition I'd consider is adding an information that when your config's cluster has a path to print an information suggesting to use this flag. But that's rather a small nit.
/lgtm
/approve

/assign @lavalamp 
for apimachinery bits approval

**Comment by FabianKramm** [comment]
@soltysh thanks for the review, I added a warning that shows up if a server path is detected and the flag is not enabled

File: staging/src/k8s.io/kubectl/pkg/cmd/proxy/proxy.go [hunk 9]
```diff
@@ -157,6 +161,15 @@ func (o *ProxyOptions) Complete(f cmdutil.Factory) error {
 		o.apiPrefix += "/"
 	}
 
+	if o.appendServerPath == false {
+		target, err := url.Parse(clientConfig.Host)
+		if err != nil {
+			return err
+		}
+		if target.Path != "" && target.Path != "/" {
+			klog.Warning("Your kube context contains a server path " + target.Path + ", use --append-server-path to automatically append the path to each request")
```
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/proxy/proxy.go:170]: :+1:

**Review by soltysh — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go [hunk 10]
```diff
@@ -246,6 +247,20 @@ func TestServeHTTP(t *testing.T) {
 			expectedPath:    "/some/path",
 			useLocationHost: true,
 		},
+		{
+			name:               "append server path to request path",
+			method:             "GET",
+			requestPath:        "/base",
+			expectedPath:       "/base/base",
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go:254]: Add a test where this expected path ends in a `/`, so that if someone "helpfully" replaces singleJoiningSlash with path.Join, they will break a test.

**Comment by lavalamp** [comment]
Needs one more test, then I can approve for api machinery.

**Comment by FabianKramm** [comment]
@lavalamp I added the test like you requested

**Comment by lavalamp** [comment]
/approve

for api machinery code.
